### DISCUSSION TITLE
Nested Annotated with mapped_column

### DISCUSSION THREAD

**Discussion by wallseat** [sqlalchemy/sqlalchemy#9729] [Ideas]
Hi, i have tried to use nested Annotated like this
```python
import uuid
from typing import Annotated
from sqlalchemy import UUID
from sqlalchemy.orm import mapped_column

uuid_ = Annotated[uuid.UUID, mapped_column(UUID(as_uuid=True))]
uuid_pk = Annotated[uuid_, mapped_column(primary_key=True, default=uuid.uuid4)]
```

but when i tried to use uuid_pk in model like this (Base inherit MappedAsDataclass)
```python3
class TestModel(Base):
    __tablename__ = "test_model"
    
    id_: Mapped[uuid_pk] = mapped_column(init=False)
    
    uuid_field: Mapped[uuid_]

test_obj = TestModel(uuid.uuid4())
```

i got an error
```
sqlalchemy.exc.ArgumentError: Mapper Mapper[TestModel(test_model)] could not assemble any primary key columns for mapped table 'test_model'
```
Does sqlalchemy support nested Annotated?

**Comment by CaselIT** [comment]
Hi

I don't think this use case has ever come up, so it seems it isn't supported at the moment

**Reply by wallseat** [reply]
I think it can be useful, Annotated support nesting (by flatting).

It seems to me like any other mapped_column should add new params that are not presented in the previous or override if presented.

**Comment by zzzeek** [comment]
no this is not something we have implemented.    (all of this functionality was just conceived in the past 8 months as it is)

I can see the thing you're going for would involve merging any number of mapped_column() constructs upwards.   Just getting the single "merge" that this does right now was really difficult...multi merges, I guess we'd just point the "merge A and B" logic at each pair and call it a day, but overall I think it's unlikely a lot of people are going to be doing what you have there (most people dont have any idea what Annotated is in the first place) so I'd be a bit concerned about the complexity / use ratio here as far as maintaining this, but I could be wrong (maybe it's easy to implement and not a lot of curveballs come up).

**Reply by wallseat** [reply]
Understood, thanks.

**Reply by zzzeek** [reply]
dont get me wrong this is certainly in the realm of doable!   i can show you where it happens if you want to look at it

**Reply by wallseat** [reply]
yes, I would be interested to see where exactly the processing of Mapped attributes takes place:)

**Reply by zzzeek** [reply]
OK the whole thing happens here, where you can see the "ispep593" check:


https://github.com/sqlalchemy/sqlalchemy/blob/228490ead7048f2e558c25b0f055bdb952272ec4/lib/sqlalchemy/orm/properties.py#L720-L808

you can see that calls into the "merge" routine which is at

https://github.com/sqlalchemy/sqlalchemy/blob/228490ead7048f2e558c25b0f055bdb952272ec4/lib/sqlalchemy/sql/schema.py#L2520-L2588

**Reply by wallseat** [reply]
https://github.com/sqlalchemy/sqlalchemy/blob/228490ead7048f2e558c25b0f055bdb952272ec4/lib/sqlalchemy/orm/properties.py#L761-L764

as far as I understand, here we just need not to exit the loop, but update the first occurrence with subsequent ones
```python
if isinstance(elem, MappedColumn):
    if not use_args_from:
        use_args_from = elem
    else:
        elem.column._merge(use_args_from.column)
```

this fixed problem with pk (and as i expect all problems with column attrs), but now i trying to find a place where Field attrs are merging... (like when we pass default_factory attr into mapped_column inside Annotated and add init=False in model)
i think we need to make the same logic like in colum merge. If present (not NO_ARG) do nothing, else use other.arg


I apologize in advance if I am wrong and didn't understand something

**Reply by zzzeek** [reply]
well we're merging an incoming mapped_column() into this one...so somehow do that recursively here.

**Reply by CaselIT** [reply]
We should probably also ensure what we are unrolling the annotation in the correct order, meaning that if two annotations have conflicting parts, the later one is used.

**Comment by wallseat** [comment]
@CaselIT 
Yes, I agree with this strategy, but right now the strategy of column._merge is `a or b`.

@zzzeek 
I'm a little stuck. I found some strange behaviour.
When i use flat Annotaned with default_factory=lambda: ... 
```python
string36_pk_auto_2 = Annotated[
    str,
    mapped_column(
        String(50),
        primary_key=True,
        nullable=False,
        default_factory=lambda: uuid.uuid4().hex[::-1],
    ),
]

class TestAuto(Base):
    __tablename__ = "test_auto"

    id_: Mapped[string36_pk_auto_2] = mapped_column(init=False)
    other_str: Mapped[string36]
```
I got an error.
`sqlalchemy.exc.IntegrityError: (sqlite3.IntegrityError) NOT NULL constraint failed: test_auto.id_`
it is only works if I use default (because sqlachemy handles it using ColumnDefault)

i remove `mapped_column(init=False)` and it worked
next, i realized smth like merging of _AttributeOptions like this
```python3
attr_options = self._attribute_options._asdict()
def _merge_attr_options(other: _AttributeOptions):
    for k, v in other._asdict().items():
        if attr_options[k] is _NoArg.NO_ARG:
            attr_options[k] = v

for elem in pep_593_components[1:]:
    if isinstance(elem, MappedColumn):
        if not use_args_from:
            use_args_from = elem
        else:
            elem.column._merge(use_args_from.column)
            
        _merge_attr_options(elem._attribute_options)
        
self._attribute_options = _AttributeOptions(**attr_options)
```
and my debugger says _attribute_options was correctly applied
![image](https://user-images.githubusercontent.com/26143672/235525830-05622ad6-b452-4122-93b7-2ecf5ca801a2.png)
But i still getting error. I think it is because all dataclass magic proccesed before? (maybe in _scan_attributes or _setup_dataclasses_transforms, I didn't have enough time to debug)

After i decided to see the order in which mapped_column will be applied with new merges.
when i use mapped_column as attr + Mapped[Annotated[...]] the order was 
```
attr: Mapped[Annotated[mc1, mc2, mc3]] = mc4

mc4 | m1 | mc2 | mc3
```
(as expected?)

but when i tried it without attr, i got this order
```
attr: Mapped[Annotated[mc1, mc2, mc3]]

mc3 | mc1 | mc2
```
and i found this rows
https://github.com/sqlalchemy/sqlalchemy/blob/main/lib/sqlalchemy/orm/decl_base.py#L1303-L1317

here we use the last one mapped_column in Annotated...

I think we need to do all the magic with the fields for dataclases somewhere else. Could you show me where it happens so that I can debug and add logic more easily?

**Reply by zzzeek** [reply]
well this is all something else, which is if the Annotated thing works with regular dataclasses, which I dont think it does at all.   Even if we made it work here, typing tools will never see it because this is not part of pep-681, and that would still be a bug.  I was told very explicitly, directly, repeatedly, by the author(s) of pep-681, that pep-681 does not in any way encourage any deviance whatsoever from the way Python stdlib dataclasses work, that they wanted to make this API in such a way that it very much *discourages* anyone from deviating from exactly what stdlib dataclasses do.  if you can maybe pick up on how ....*emphatic*, they were about this.

Since dataclasses themselves dont support this kind of thing with Field(), so I wasn't worrying about making this very new, very experimental "Annotated" thing work at the same time with this other very new, very experimental "mapped classes are directly dataclasses" thing.

next message...

**Reply by zzzeek** [reply]
that all said if we wanted to transfer dataclass attributes from Annotated, I would not change the current logic as I'd rather not change it very much, just add a second scan step to MappedColumn, have _AttributeOptions._get_arguments_for_make_dataclass call an additional hook on _DCAttributeOptions that can do the same traversal into each Annotated and find only dataclass attributes.

**Reply by zzzeek** [reply]
if we *dont* do that, then the DC attributes inside of `Annotated[..., mapped_column()]` should probably raise, not silently fail.

**Reply by CaselIT** [reply]
> if we _dont_ do that, then the DC attributes inside of `Annotated[..., mapped_column()]` should probably raise, not silently fail.

100% agree. a raise is better that silently ignoring something

**Reply by wallseat** [reply]
So, I read this and...
I don't think that it necessary adding logic around these fields now, because
1. They will still not be recognized by typecheckers anywhere other than the class itself.
2. I'm not sure that i will have enough time to do smth like this (in the near future)
  > that all said if we wanted to transfer dataclass attributes from Annotated, I would not change the current logic as I'd rather not change 
     it very much, just add a second scan step to MappedColumn, have _AttributeOptions._get_arguments_for_make_dataclass call an 
     additional hook on _DCAttributeOptions that can do the same traversal into each Annotated and find only dataclass attributes.
     
but it looks like some kind of "undefined" behavior and it is worth at least somehow limiting it. I also agree that it's necessary to raise if those parameters are defined as Mapped[Annotated[...]]. So i think, we can prohibit such use and write why such a decision was made, and later, if possible, solve this problem...
